cluster/kubernetes/kubeyaml.go

```diff
@@ -0,0 +1,47 @@
+package kubernetes
+
+import (
+	"bytes"
+	"errors"
+	"os"
+	"os/exec"
+	"strings"
+)
+
+// KubeYAML is a placeholder value for calling the helper executable
+// `kubeyaml`.
+type KubeYAML struct {
+}
+
+// Image calls the kubeyaml subcommand `image` with the arguments given.
+func (k KubeYAML) Image(in []byte, ns, kind, name, container, image string) ([]byte, error) {
+	args := []string{"image", "--namespace", ns, "--kind", kind, "--name", name}
+	args = append(args, "--container", container, "--image", image)
+	return execKubeyaml(in, args)
+}
+
+// Annotate calls the kubeyaml subcommand `annotate` with the arguments as given.
+func (k KubeYAML) Annotate(in []byte, ns, kind, name string, policies ...string) ([]byte, error) {
+	args := []string{"annotate", "--namespace", ns, "--kind", kind, "--name", name}
+	args = append(args, policies...)
+	return execKubeyaml(in, args)
+}
+
+func execKubeyaml(in []byte, args []string) ([]byte, error) {
```

Comment:
Consume an `io.Reader` here! That way we don't have to read in a yaml's `os.File` (which implements `io.Reader`) into a `[]byte` only to wrap it back up in a `bytes.NewBuffer` later.

We only need to go down to the byte-slice level if we actually want to look at what's in there. Once our flux code stops modifying yaml directly, it'll only be doing two things:
- reading in yaml for the sake of parsing (but not editing), and
- forwarding yaml to kubeyaml,
neither of which should actually require us to know anything more than "we have a source of bytes"

Steve Francia says it better than I might hope to: https://youtu.be/29LLRKIL_TI?t=199

Reply:
Produce a reader too, maybe

Reply:
> Consume an io.Reader here!

Great idea.

> Produce a reader too, maybe

Maybe; we will almost always want to write the result back to the file, and returning a reader makes this a bit harder.

Reply:
The only counter-argument I can summon is that we read the input more than once, at least for annotations, so this could create more faff rather than reducing it.

Reply:
> Maybe; we will almost always want to write the result back to the file, and returning a reader makes this a bit harder.

How so? Given a reader, `r`, and an open file, `f`, `io.Copy(f, r)` will write out the file.


> we read the input more than once, at least for annotations

For now! I can't see a reason why in future we would need to read the file more than once specifically for annotations. One call to `YAML.Unmarshal(...)` will extract all the info we need, no?

Reply:
> One call to YAML.Unmarshal(...) will extract all the info we need, no?

It does; _then_ we want to pass the input to `kubeyaml`.

Reply:
Sounds like we want to use [`io.TeeReader`](https://godoc.org/io#TeeReader)

Reply:
I like this suggestion overall, and I've started implementing it in a branch (`refactor/bytes-to-readers`). The consequences are quite far-ranging though, so best to do in another PR after the fact, I think.

Reply:
I agree; a separate PR seems the best way to go.